### PR TITLE
[SPARK-14388][SQL] Implement CREATE TABLE

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?

This patch implements the `CREATE TABLE` command using the `SessionCatalog`. Previously we handled only `CTAS` and `CREATE TABLE ... USING`. This requires us to refactor `CatalogTable` to accept various fields (e.g. bucket and skew columns) and pass them to Hive.

WIP: Note that I haven't verified whether this actually works yet! But I believe it does.
## How was this patch tested?

Tests will come in a future commit.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 1]
```diff
@@ -176,29 +176,6 @@ case class DescribeDatabase(
...
 /**
- * A command that renames a table/view.
- *
- * The syntax of this command is:
- * {{{
- *    ALTER TABLE table1 RENAME TO table2;
- *    ALTER VIEW view1 RENAME TO view2;
- * }}}
- */
-case class AlterTableRename(
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: moved to `tables.scala`. I just moved this one command for now to avoid inflating the diff too much.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 2]
```diff
@@ -320,8 +320,6 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
     assertUnsupported("ALTER TABLE dbx.tab1 NOT STORED AS DIRECTORIES")
   }
 
-  // TODO: ADD a testcase for Drop Database in Restric when we can create tables in SQLContext
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: this TODO comment is duplicated in line 197

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 3]
```diff
@@ -271,8 +271,7 @@ createFileFormat
     ;
 
 fileFormat
-    : INPUTFORMAT inFmt=STRING OUTPUTFORMAT outFmt=STRING (SERDE serdeCls=STRING)?
-      (INPUTDRIVER inDriver=STRING OUTPUTDRIVER outDriver=STRING)?                         #tableFileFormat
```
Comment (andrewor14) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @hvanhovell I deleted this because Hive doesn't support it. Why was this added in the first place? Is there any supporting documentation for this somewhere?

Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @andrewor14 I wanted to make sure we supported the same grammar as Hive and I used their grammars as a basis. So this is defined in the following two locations:

- https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g#L1335-L1336
- https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g#L1335-L1336

The main idea was that I could throw better errors. But if it is not supported by Hive itself then please remove it!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 4]
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
     tableType: CatalogTableType,
     storage: CatalogStorageFormat,
     schema: Seq[CatalogColumn],
-    partitionColumns: Seq[CatalogColumn] = Seq.empty,
-    sortColumns: Seq[CatalogColumn] = Seq.empty,
+    partitionColumnNames: Seq[String] = Seq.empty,
```
Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: The hive syntax has PARTITIONED BY (col_name  data_type  [COMMENT col_comment], ...). The column definition here is the same as the ones in schema. CatalogColumn can represent these 3 aspects of each column.. why we need to change to use String here?

**Comment by gatorsmile** [comment]
@andrewor14 Just want to let you know, @xwu0226 is doing the command `SHOW CREATE TABLE`. He is writing many test cases in the PR: https://github.com/apache/spark/pull/12132/files. It might help you in this PR. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ * This is not intended for temporary tables.
+ *
+ * The syntax of using this command in SQL is:
+ * {{{
+ *   CREATE [EXTERNAL] TABLE [IF NOT EXISTS] [db_name.]table_name
+ *   [(col1 data_type [COMMENT col_comment], ...)]
+ *   [COMMENT table_comment]
+ *   [PARTITIONED BY (col3 data_type [COMMENT col_comment], ...)]
+ *   [CLUSTERED BY (col1, ...) [SORTED BY (col1 [ASC|DESC], ...)] INTO num_buckets BUCKETS]
+ *   [SKEWED BY (col1, col2, ...) ON ((col_value, col_value, ...), ...)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: @andrewor14 I just did a quick check with our `InsertIntoHiveTable` command. Seems this command does not really understand how to handle a table having specifications on `CLUSTERED BY`, `SORTED BY` or `SKEWED BY`. How about we just throw exceptions when a define provide these specs?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 3 above)
Comment (yhuai) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Let's remove it. I have never seen this before.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 6]
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
...
+    viewText: Option[String] = None,
+    comment: Option[String] = None) {
+
+  // Verify that the provided columns are part of the schema
+  private val colNames = schema.map(_.name).toSet
+  private def requireSubsetOfSchema(cols: Seq[String], colType: String): Unit = {
+    require(cols.toSet.subsetOf(colNames), s"$colType columns (${cols.mkString(", ")}) " +
+      s"must be a subset of schema (${colNames.mkString(", ")}) in table '$identifier'")
+  }
+  requireSubsetOfSchema(partitionColumnNames, "partition")
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: I find the following example from Hive's doc
```
CREATE TABLE page_view(viewTime INT, userid BIGINT,
     page_url STRING, referrer_url STRING,
     ip STRING COMMENT 'IP Address of the User')
 COMMENT 'This is the page view table'
 PARTITIONED BY(dt STRING, country STRING)
 ROW FORMAT DELIMITED
   FIELDS TERMINATED BY '\001'
STORED AS SEQUENCEFILE;
```

Will this CREATE TABLE statement trigger an error at here?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 7]
```diff
@@ -665,9 +672,17 @@ private[hive] class HiveClientImpl(
       case CatalogTableType.INDEX_TABLE => HiveTableType.INDEX_TABLE
       case CatalogTableType.VIRTUAL_VIEW => HiveTableType.VIRTUAL_VIEW
     })
-    hiveTable.setFields(table.schema.map(toHiveColumn).asJava)
-    hiveTable.setPartCols(table.partitionColumns.map(toHiveColumn).asJava)
+    // TODO: pass the comment to Hive somehow
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: comment is stored in the table properties with a field called `comment`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 6 above)
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: no because we put the partition columns in the schema when we parse it. There's a comment in `HiveSqlParser` that describes this.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 7 above)
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I see. That's what I suspected

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 8]
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
+      case (fileFormat, null) if fileFormat != null =>
+        fileFormat match {
+          // Expected format: INPUTFORMAT input_format OUTPUTFORMAT output_format
+          case c: TableFileFormatContext => visitTableFileFormat(c)
+          // Expected format: SEQUENCEFILE | TEXTFILE | RCFILE | ORC | PARQUET | AVRO
+          case c: GenericFileFormatContext => visitGenericFileFormat(c)
+        }
+      case (null, storageHandler) if storageHandler != null =>
+        throw new ParseException("Operation not allowed: ... STORED BY storage_handler ...", ctx)
+      case (null, null) =>
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: I don't think this is possible. If they are both empty, the `CreateFileFormatContext` should be null to begin with.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 9]
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
-   * Create a [[CatalogStorageFormat]]. This is part of the [[CreateTableAsSelect]] command.
+   * Create a [[CatalogStorageFormat]] for creating tables.
    */
   override def visitCreateFileFormat(
       ctx: CreateFileFormatContext): CatalogStorageFormat = withOrigin(ctx) {
-    if (ctx.storageHandler == null) {
-      typedVisit[CatalogStorageFormat](ctx.fileFormat)
-    } else {
-      visitStorageHandler(ctx.storageHandler)
+    (ctx.fileFormat, ctx.storageHandler) match {
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: The parser either allows a fileFormat or a storageHandler here (not both). So we could simplify this.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 10]
```diff
@@ -397,11 +429,28 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
+   *   [MAP KEYS TERMINATED BY char]
+   *   [LINES TERMINATED BY char]
+   *   [NULL DEFINED AS char]
+   * }}}
    */
-  override def visitStorageHandler(
-      ctx: StorageHandlerContext): CatalogStorageFormat = withOrigin(ctx) {
-    throw new ParseException("Storage Handlers are currently unsupported.", ctx)
+  private def visitRowFormat(ctx: RowFormatContext): CatalogStorageFormat = withOrigin(ctx) {
+    ctx match {
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Minor: You could also call `ctx.accept(this)` here and cast it, but I get that this is more typed and less opaque.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 11]
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
+   *   [SKEWED BY (col1, col2, ...) ON ((col_value, col_value, ...), ...)
+   *   [STORED AS DIRECTORIES]
+   *   [ROW FORMAT row_format]
+   *   [STORED AS file_format | STORED BY storage_handler_class [WITH SERDEPROPERTIES (...)]]
+   *   [LOCATION path]
+   *   [TBLPROPERTIES (property_name=property_value, ...)]
+   *   [AS select_statement];
+   * }}}
    */
   override def visitCreateTable(ctx: CreateTableContext): LogicalPlan = {
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Could you add withOrigin here? I apparently forgot to...

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 12]
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
-    if (ctx.query == null) {
-      HiveNativeCommand(command(ctx))
+    val (name, temp, ifNotExists, external) = visitCreateTableHeader(ctx.createTableHeader)
+    // TODO: implement temporary tables
+    if (temp) {
+      throw new AnalysisException(
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: I would throw a ParseException here. The advantage is that the error reporting of these exceptions is a bit better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 13]
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
     tableType: CatalogTableType,
     storage: CatalogStorageFormat,
     schema: Seq[CatalogColumn],
-    partitionColumns: Seq[CatalogColumn] = Seq.empty,
-    sortColumns: Seq[CatalogColumn] = Seq.empty,
+    partitionColumnNames: Seq[String] = Seq.empty,
+    sortColumnNames: Seq[String] = Seq.empty,
```
Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: How do we maintain the ordering of the sort columns here? For the clause SORTED BY (col_name [ASC|DESC], ...), within CLUSTERED BY clause.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 8 above)
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: sure

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 10 above)
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: yeah, that's the whole reason why I changed it. There's too much magic going on in `typedAccept` or whatever it's called, especially in code that is generated and hidden from the IDE.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 12 above)
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: I'll address all of these in SPARK-14414. For now I just made this (and a few others) an `AnalysisException` so we can reuse some code in testing. I'll fix this before 2.0.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 13 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: I think we do not support SORTED BY and CLUSTERED BY for now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 14]
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
...
+    // e.g. ('2008-08-08', 'us), ('2009-09-09', 'uk')
+    skewColumnValues: Seq[Seq[String]] = Seq.empty,
     numBuckets: Int = 0,
     createTime: Long = System.currentTimeMillis,
     lastAccessTime: Long = System.currentTimeMillis,
     properties: Map[String, String] = Map.empty,
     viewOriginalText: Option[String] = None,
-    viewText: Option[String] = None) {
+    viewText: Option[String] = None,
+    comment: Option[String] = None) {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Maybe renaming it to `description` (be consistent with `CatalogDatabase`)?

Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Does this "comment" mean the comment for the table, as defined in the syntax `create table t1 (c1 int) COMMENT 'abc'`?  If it is, shall it be put into `properties`, it seems Hive puts this table comment into TBLPROPERTIES. so it is consistent when creating hive metastore table from CatalogTable object?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 13 above)
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: also that has nothing to do with this patch

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 14 above)
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: I actually prefer to keep it as comment because it's `COMMENT` in the query and `"comment"` in Hive's table properties

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 5 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: ok

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 11 above)
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 14 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: ok

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 15]
```diff
@@ -121,84 +123,114 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
+   * skewed columns and storage handlers (STORED BY).
+   *
+   * Expected format:
+   * {{{
+   *   CREATE [TEMPORARY] [EXTERNAL] TABLE [IF NOT EXISTS] [db_name.]table_name
+   *   [(col1 data_type [COMMENT col_comment], ...)]
+   *   [COMMENT table_comment]
+   *   [PARTITIONED BY (col3 data_type [COMMENT col_comment], ...)]
+   *   [CLUSTERED BY (col1, ...) [SORTED BY (col1 [ASC|DESC], ...)] INTO num_buckets BUCKETS]
+   *   [SKEWED BY (col1, col2, ...) ON ((col_value, col_value, ...), ...) [STORED AS DIRECTORIES]]
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Remove this line while resolving the conflicts?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 6 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 16]
```diff
@@ -121,84 +123,114 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
+   *   [TBLPROPERTIES (property_name=property_value, ...)]
+   *   [AS select_statement];
+   * }}}
    */
-  override def visitCreateTable(ctx: CreateTableContext): LogicalPlan = {
-    if (ctx.query == null) {
-      HiveNativeCommand(command(ctx))
+  override def visitCreateTable(ctx: CreateTableContext): LogicalPlan = withOrigin(ctx) {
+    val (name, temp, ifNotExists, external) = visitCreateTableHeader(ctx.createTableHeader)
+    // TODO: implement temporary tables
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: +1

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 17]
```diff
@@ -121,84 +123,114 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
+    val rowStorage = Option(ctx.rowFormat).map(visitRowFormat).getOrElse(EmptyStorageFormat)
+    val location = Option(ctx.locationSpec).map(visitLocationSpec)
+    val storage = CatalogStorageFormat(
+      locationUri = location,
+      inputFormat = fileStorage.inputFormat.orElse(defaultStorage.inputFormat),
+      outputFormat = fileStorage.outputFormat.orElse(defaultStorage.outputFormat),
+      serde = rowStorage.serde.orElse(fileStorage.serde).orElse(defaultStorage.serde),
+      serdeProperties = rowStorage.serdeProperties ++ fileStorage.serdeProperties)
+
+    // TODO support the sql text - have a proper location for this!
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Maybe explain this?

**Comment by yhuai** [comment]
LGTM

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 17 above)
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: @hvanhovell maybe explain this? :)

**Comment by andrewor14** [comment]
Thanks for the reviews. Latest commit should pass tests.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 12 above)
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: (I actually fixed this in this patch, but we still need to fix it in all the other places in SPARK-14414)

**Comment by yhuai** [comment]
test this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 17 above)
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: In the old situation we were storing the `comment` into the `viewText` column which didn't seem right. I think we can remove this one now that we actually have a comment field.

**Comment by hvanhovell** [comment]
LGTM - pending jenkins

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
I've ignored some tests in `HiveCompatibilitySuite` for now. I'll have a look at them shortly. I just wanted to see if the rest of the tests will pass.

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala [hunk 18]
```diff
@@ -509,7 +555,7 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     "date_comparison",
     "date_join1",
     "date_serde",
-    "decimal_1",
+    //"decimal_1", // TODO: cannot parse column decimal(5)
```
Comment (yhuai) [inline sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: Yea. We should support decimal(5). At here, the scale is 0.

Comment (yhuai) [inline sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: https://issues.apache.org/jira/browse/SPARK-14591

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala [hunk 19]
```diff
@@ -490,15 +538,13 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     "count",
     "cp_mj_rc",
     "create_insert_outputformat",
-    "create_like_tbl_props",
+    //"create_like_tbl_props", // TODO: results don't match
```
Comment (yhuai) [inline sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: https://issues.apache.org/jira/browse/SPARK-14592

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala [hunk 20]
```diff
@@ -790,14 +821,13 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
...
     "orc_dictionary_threshold",
-    "orc_empty_files",
     "order",
     "order2",
     "outer_join_ppr",
     "parallel",
     "parenthesis_star_by",
     "part_inherit_tbl_props",
-    "part_inherit_tbl_props_empty",
+    //"part_inherit_tbl_props_empty", // TODO: results don't match
```
Comment (yhuai) [inline sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: This one and others that are commented out below are caused by differences of describe table. There are mainly three differences. First, for managed tables, we will EXTERNAL set to false in table properties. For tables that are not bucketed, we will have numBuckets set to 0 instead of -1. For tables that does not specify the file format, the output format will be org.apache.hadoop.hive.ql.io.IgnoreKeyTextOutputFormat instead of org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat (For Hive, it somehow replaced even if the explain output still shows IgnoreKeyTextOutputFormat). We do not need to block this PR for any of them.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 21]
```diff
@@ -121,84 +123,115 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
-      )
+      val defaultHiveSerde = HiveSerDe.sourceToSerDe(defaultStorageType, hiveConf)
+      CatalogStorageFormat(
+        locationUri = None,
+        inputFormat = defaultHiveSerde.flatMap(_.inputFormat)
+          .orElse(Some("org.apache.hadoop.mapred.TextInputFormat")),
+        outputFormat = defaultHiveSerde.flatMap(_.outputFormat)
+          .orElse(Some("org.apache.hadoop.hive.ql.io.IgnoreKeyTextOutputFormat")),
+        serde = defaultHiveSerde.flatMap(_.serde)
+          .orElse(Some("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe")),
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: This one breaks the CTAS test.

Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: I guess we should fill this in as late as possible? Probably in `toHiveTable`? In `HiveClientImpl.toHiveTable`, we first create a `org.apache.hadoop.hive.ql.metadata.Table`, which uses `MetadataTypedColumnsetSerDe` as the default one (so, originally `InsertIntoHiveTableSuite` was broken without this change).

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 22]
```diff
@@ -435,13 +480,15 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
     ctx.colType.asScala.map { col =>
       CatalogColumn(
-        formatter(col.identifier.getText),
-        col.dataType.getText.toLowerCase, // TODO validate this?
+        col.identifier.getText.toLowerCase,
+        // Note: for types like "STRUCT<myFirstName: STRING, myLastName: STRING>" we can't
+        // just convert the whole type string to lower case, otherwise the struct field names
+        // will no longer be case sensitive. Instead, we rely on our parser to get the proper
+        // case before passing it to Hive.
+        HiveMetastoreTypes.toDataType(col.dataType.getText).simpleString,
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: I think we need to use `parseDataType`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 23]
```diff
@@ -488,7 +488,7 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
         // just convert the whole type string to lower case, otherwise the struct field names
         // will no longer be case sensitive. Instead, we rely on our parser to get the proper
         // case before passing it to Hive.
-        HiveMetastoreTypes.toDataType(col.dataType.getText).simpleString,
+        CatalystSqlParser.parseDataType(col.dataType.getText).simpleString,
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: MINOR/NIT: The DataType parsing is done in the `AstBuilder`, so we really don't need to parse this string again. You could use (magical/evil) `typedVisit[DataType](col.dataType)` here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 24]
```diff
@@ -435,13 +480,15 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
     ctx.colType.asScala.map { col =>
       CatalogColumn(
-        formatter(col.identifier.getText),
-        col.dataType.getText.toLowerCase, // TODO validate this?
+        col.identifier.getText.toLowerCase,
+        // Note: for types like "STRUCT<myFirstName: STRING, myLastName: STRING>" we can't
+        // just convert the whole type string to lower case, otherwise the struct field names
+        // will no longer be case sensitive. Instead, we rely on our parser to get the proper
+        // case before passing it to Hive.
+        CatalystSqlParser.parseDataType(col.dataType.getText).simpleString,
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: MINOR/NIT: The DataType parsing is done in the `AstBuilder`, so we really don't need to parse this string again. You could use (magical/evil) `typedVisit[DataType](col.dataType)` here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 25]
```diff
@@ -675,34 +680,38 @@ private[hive] class HiveClientImpl(
...
     hiveTable.setOwner(conf.getUser)
     hiveTable.setNumBuckets(table.numBuckets)
     hiveTable.setCreateTime((table.createTime / 1000).toInt)
     hiveTable.setLastAccessTime((table.lastAccessTime / 1000).toInt)
     table.storage.locationUri.foreach { loc => shim.setDataLocation(hiveTable, loc) }
     table.storage.inputFormat.map(toInputFormat).foreach(hiveTable.setInputFormatClass)
     table.storage.outputFormat.map(toOutputFormat).foreach(hiveTable.setOutputFormatClass)
-    table.storage.serde.foreach(hiveTable.setSerializationLib)
+    hiveTable.setSerializationLib(
+      table.storage.serde.getOrElse("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe"))
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: hmm... Without this PR, the storage related fields of the table defined in `org.apache.spark.sql.hive.MetastoreDataSourcesSuite.persistent JSON table` are:
```
[SerDe Library:      	org.apache.hadoop.hive.serde2.MetadataTypedColumnsetSerDe	 ]
[InputFormat:        	org.apache.hadoop.mapred.SequenceFileInputFormat	 ]
[OutputFormat:       	org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat	 ]
```

Looks like the change at here somehow breaks the test (without the change, we can somehow trigger a weird code path in Hive and get at least have one column called `col`).

Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: For now, if the schema does not have any field, can we set it to the following one?
```
[# col_name            	data_type           	comment             ]
[	 	 ]
[col                 	array<string>       	from deserializer   ]
```

So, we try to preserve the existing (and weird) behavior.

Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I submitted PR to try it out (https://github.com/apache/spark/pull/12363). You can find my change in the last commit (https://github.com/apache/spark/pull/12363/commits/ab70cb751cce8ca2e0757b9fa523534207864328).

Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: why are there all these undocumented implicit behaviors in Hive :(

**Comment by andrewor14** [comment]
OK, I cherry-picked your changes and updated the comment.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 25 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Now, after a few bad experiences, I think we need to add all the test cases to ensure the Hive APIs work as what we expected. See another issue I hit in `alter table ... drop partition`: https://github.com/apache/spark/pull/12220#discussion-diff-59555113

Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I feel it is better to have more checks in SessionCatalog to make sure that a request is valid (it can help your case).

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Yeah, agree if you are talking about my case. : )

**Comment by yhuai** [comment]
OK. Merging to master.
